trollimage/colormap.py

```diff
@@ -88,28 +101,37 @@ def _colorize(arr, colors, values):
 
 def _interpolate_rgb_colors(arr, colors, values):
     interp_xp_coords = np.array(values)
-    hcl_colors = _convert_rgb_list_to_hcl(colors)
-    interp_y_coords = np.array(hcl_colors)
+    interp_y_coords = rgb2hcl(colors)
     if values[0] > values[-1]:
         # monotonically decreasing
         interp_xp_coords = interp_xp_coords[::-1]
         interp_y_coords = interp_y_coords[::-1]
-    channels = [np.interp(arr,
-                          interp_xp_coords,
-                          interp_y_coords[:, i])
-                for i in range(3)]
-    channels = list(hcl2rgb(*channels))
-    return channels
+    # handling NaN Hue values for interpolation
+    new_hues = interp_y_coords[..., 0]
+    if np.isnan(new_hues).all():
+        new_hues[..., :] = 0
+    else:
+        if np.isnan(new_hues[..., 0]):
+            new_hues[..., 0] = new_hues[~np.isnan(new_hues)][0]
+        while np.isnan(new_hues).any():
+            new_hues[..., 1:] = np.where(np.isnan(new_hues[..., 1:]), new_hues[..., :-1], new_hues[..., 1:])
+    interp_y_coords[..., 0] = np.unwrap(new_hues)
+    interp_hcl = np.zeros(arr.shape + (3,), dtype=interp_y_coords.dtype)
+    interp_hcl[..., 0] = np.interp(arr, interp_xp_coords, interp_y_coords[..., 0])
```

Comment:
I think I see what you are doing here to fill nans, taking the "left" neighbour, right?
But why not just use interpolation for this, eg:
```suggestion
        interp_xp_coords_hues = interp_xp_coords[~np.isnan(new_hues)]
        new_hues = new_hues[~np.isnan(new_hues)]  # only valid hues
    new_hues = np.unwrap(new_hues)
    interp_hcl = np.zeros(arr.shape + (3,), dtype=interp_y_coords.dtype)
    interp_hcl[..., 0] = np.interp(arr, interp_xp_coords_hues, new_hues, left=new_hues[..., 0], right=new_hues[..., -1])
```

Reply:
Here is a comparison of the effect of this for rdbu.
This PR:
![rdbu_pr](https://user-images.githubusercontent.com/167802/219330886-37127512-567f-44d4-a0a3-a283b36c8651.png)
With the suggested change:
![rdbu_pr_mod](https://user-images.githubusercontent.com/167802/219330983-3caec02d-91cd-4ab2-b5a1-f42ec2e608f5.png)

Reply:
Here is for the rdbu_simple I introduced in a previous example.
This PR:
![rdbu_simple_pr](https://user-images.githubusercontent.com/167802/219332243-1cb7ab15-ce1c-4a58-9bf3-149d4a1749b8.png)
With the suggested change:
![rdbu_simple_pr_mod](https://user-images.githubusercontent.com/167802/219332269-97f46eae-23f8-4023-9c6f-c36de3622dba.png)

Looks like the suggested change keeps better the perceptual gradients.

Reply:
Oh I think there is a misunderstanding here. The NaN handling wasn't to remove the colors/elements with the NaNs in the Hue, that would remove entire colors from the colormap. It was meant to handle the case of grayscale colors where Hue results in NaN (done on purpose by me) and we need to take the previous hue. For example, if you had the color green which is a hue of like pi/3 and you needed to interpolate to the next color of white, the basic math of the algorithm could/would result in white having a hue of 0 (as a fallback for a `/ 0`). But we don't want hue to be interpolated between pi/3 and 0, we want hue to stay at pi/3 because when chroma is 0 hue has no effect.

The problem with my current approach though is that if we then interpolate from that white to blue (for example), we'd still be interpolating hue from pi/3 to pi instead of immediately jumping to pi. My guess/hope is that this happens for any grayscale color and that I can fix it. It may not help with the luminance/chroma issues you identified, but maybe it can clean up a couple other things.

Side note: I think your `left` and `right` keyword arguments match the default behavior of interp don't they?

Reply:
Yes, I gathered that it was for handling monochromatic situations. But I hadn't understood that C and l were nans too then...

Reply:
I see now that you're only dropping the Nan elements in hue. One problem with this is it gets rid of the behavior of the control points. If the white hue was not perfectly between the two end colors then the hue interpolation between the two colors wouldn't be weighted properly. Like the middle color being at 0.4, we need to interpolate first to the 0.4 spot and then interpolate "more slowly" from 0.4 to 1.0...at least I think. Regardless, I don't think we want to interpolate hue in this case...right? Don't we want to go "straight" to the grayscale/monochromatic value and then from there (the center of the colorspace diagram) to the non-grayscale value.

Reply:
Ok I hacked a solution for the rdbu_simple example you did and here's what I get:

![rdbu_simple](https://user-images.githubusercontent.com/1828519/219447460-fff41160-b3e3-4821-9d1d-c490b0722e28.png)

Which to me looks as close to `main` as possible. I basically took the 3 HCL colors, repeated the 2 element (because it is NaN) and also repeated the `values` in the same way. Then I set the first NaN hue to hue[0] and the second NaN to hue[-1].

Reply:
Nevermind, that still looks very much like it did before but with better gradient on the blue side. The red side is still a different color I feel like.

Reply:
Regarding nans, doesn't this line show that only hues get nans? https://github.com/pytroll/trollimage/pull/122/files#diff-5f1ad72add694d03d94098c73ec8d5db5be7ef6ebb78648a606735d346769c0cR133
That would actually make sense, otherwise rdbu_simple wouldn't work at all: we need to keep the info that the middle point has 0 chroma and max luminance.

Regarding the interpolation technique:
Well, I think it depends on the situation. We can see artefacts in the diverging colormaps (also in main) because in principle we should have a discontinuity in the hue. Eg rdbu should be blue hue on the right and red on the left, and in the middle the hue has to change at exactly the middle point. Since we don't do that we get varying hues around the centre. (I intend to fix this in a future PR through the usage of `np.piecewise` or dask equivalent)

As for the usage of `interp`, you are absolutely right, the `right` and `left` parameters I proposed are redundant.